### PR TITLE
Update dependencies to latest versions

### PR DESCRIPTION
**Pull request by merlinnot**
### Description

Issue #439, partial resolution.

This PR:
- updates all dependencies to latest versions
- moves some `@types` dependencies to devDependencies
- refactors a few functions that did not compile because of improved types
- renames `typings` field to `types`, as it's the base name
- changes `node_modules/.bin/tsc` to `tsc` in all scripts in `package.json`
- orders fields in `package.json` in an alphabetical order, so it's easier to find relevant fields

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,70 +1,70 @@
...
+    "prettier": "^1.13.7",
+    "sinon": "^7.3.2",
+    "ts-node": "^8.1.0",
+    "typescript": "^3.4.0"
   },
-  "repository": {
-    "type": "git",
-    "url": "git+https://github.com/firebase/firebase-functions.git"
+  "engines": {
+    "node": ">=6.0.0"
```
Comment (mad-it) [inline package.json]: Support for engine 6.0.0 is being dropped right?

**Comment by thechenky** [comment]
I'm not a fan of alphabetized `package.json` as it was quite difficult for me to find things like `name` which I'm used to seeing at the top. Additionally, I think it's odd to have `devDependencies` and `dependencies` not next to each other. Can we leave that file as is?

**Comment by merlinnot** [comment]
I change the order of top-level properties back to original.

**Comment by merlinnot** [comment]
Closing in favour of #450

**Pull request closed without merging**